### PR TITLE
gh-93476: Fraction.limit_denominator speed increase

### LINKED ISSUE #93476: Increase speed of Fractions.limit_denominator
**Feature or enhancement**

Speed up Fractions.limit_denominator by not creating intermediary Fraction objects

**Pitch**

`Fractions.limit_denominator` returns a new Fraction, but along the way for Fractions > max_denominator creates 3 Fraction objects that are discarded: the fraction farther from the bound and two temporary fractions (`bound1 - self` and `bound2 - shelf`) which are created just for getting their absolute values.

Similar code speedups have been part of the `music21` library since 2015--where we call limit_denominator after every arithmetic operation on note placement--and timing tests give a 7.3x speedup.  (see [music21 Source](https://github.com/cuthbertLab/music21/blob/dcd23748052bfc4e02d5f94a9ff3de5b88ffcb89/music21/common/numberTools.py#L150) )

For fractions with denominators equal to or below max_denominator, instantiate the new Fraction with `Fraction(self._numerator, self._denominator)` instead of `Fraction(self)` for a 23% speedup.

PR to follow.

### PR DESCRIPTION
**Pull request by mscuthbert**
Create only one new Fraction object during `limit_denominator()` instead of the four previously made.

Fixes #93476

I believe that each of the calls to `Fraction()` can be called with `_normalize=False` but I have not included that in this PR to limit the scope of changes to the minimum to verify no change in implementation.

### CODE REVIEW

**Comment by sweeneyde** [comment]
Hi @mscuthbert, thanks for the PR.

It looks like there are some failing tests:
```
FAIL: testLimitDenominator (test.test_fractions.FractionTest.testLimitDenominator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/Lib/test/test_fractions.py", line 357, in testLimitDenominator
    self.assertEqual(F(201, 200).limit_denominator(100), F(1))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Fraction(101, 100) != Fraction(1, 1)
```

I'm guessing it's an edge case in the `difference > 0` logic.

Beyond that, I'd like to see some compelling benchmarks for this change across a range of numerators/denominators/max_denominators before deciding that it's worth the extra code. This inlining doesn't change the asymptotics on large numbers, just the constant overhead, right?

**Comment by mscuthbert** [comment]
> Hi @mscuthbert, thanks for the PR.
> 
> It looks like there are some failing tests:
> 
> ```
> FAIL: testLimitDenominator (test.test_fractions.FractionTest.testLimitDenominator)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/vsts/work/1/s/Lib/test/test_fractions.py", line 357, in testLimitDenominator
>     self.assertEqual(F(201, 200).limit_denominator(100), F(1))
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> AssertionError: Fraction(101, 100) != Fraction(1, 1)
> ```
> 
> I'm guessing it's an edge case in the `difference > 0` logic.

Thanks @sweeneyde -- I'll look into this -- I'll do a clean build and see if I missed something in the renaming to fit PEP8 variable names (I had camelCase everwhere in my original code).  I think I swapped bound1 and bound2 in the equality test and missed the edge case of both being equal.

> Beyond that, I'd like to see some compelling benchmarks for this change across a range of numerators/denominators/max_denominators before deciding that it's worth the extra code. This inlining doesn't change the asymptotics on large numbers, just the constant overhead, right?

Sure, I'll make sure to post some as soon as I have all the tests passing, and run some tests on small denominators, which I didn't do.

`__new__` on Fractions by default runs normalization code which calls math.gcd.  The gcd algorithm runs in O(log n) time, so we'll be eliminating 3 of the 4 calls to that in this code.  It doesn't reduce the algorithmic complexity of this routine, but should be a greater speed up for larger Fraction objects.

We can eliminate the fourth gcd call also if we do not need to retain behavior on an unreduced Fraction created with the private and undocumented keyword `_normalize=False` does not need to maintain backwards compatibility.  (This will speed up calling limit_denominator on Fractions whose denominators is below the limit a lot.)     One example of an existing call that does not guarantee a Fraction that is normalized if the original fraction was not in lowest terms is `__pow__` (and `__rpow__`.   For instance, where `Fraction(3, 12, _normalize=False)**2` returns `Fraction(9, 144)` and not `Fraction(1, 16)`.  Furthermore `3**Fraction(2, 1)` returns `9` (int) while `3**Fraction(4, 2, _normalize=False)` returns `9.0` (float).  Other routines that also give different results with Fractions that skipped normalization are `as_integer_ratio()` (and numerator, denominator, obviously), abs, neg, etc. so I think the general rule for all these methods is to trust that they are being called on a normalized Fraction only.  There was not a discussion about this during the addition of the `_normalize` parameter in Py 3.5: https://bugs.python.org/issue21136

**Comment by mscuthbert** [comment]
Here are the timing tests I've run.  All fractions checked that they were in lowest terms at declaration.

```
In [1]: from fractions import Fraction as F1
In [2]: from fract312 import Fraction as F2   # temp filename
In [3]: F1 is F2
Out[3]: False

In [4]: f1 = F1(1_234_567, 2_345_678)  # relatively prime
In [5]: f2 = F2(1_234_567, 2_345_678)

# 3.5x speed for typically large number, medium denominator
In [6]: %timeit f1.limit_denominator(1500)
8.95 µs ± 170 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [7]: %timeit f2.limit_denominator(1500)
2.54 µs ± 71.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# 1.33x no change to Fraction, create new one
In [8]: %timeit f1.limit_denominator(2_345_679)
1.17 µs ± 24.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
In [9]: %timeit f2.limit_denominator(2_345_679)
879 ns ± 18.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

# 3.9x limit to small denominator
In [10]: %timeit f1.limit_denominator(1)
8.67 µs ± 202 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [11]: %timeit f2.limit_denominator(1)
2.21 µs ± 81.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# small fraction
In [12]: f1 = F1(5, 4)
In [13]: f2 = F2(5, 4)

# small limit, 4.4x
In [14]: %timeit f1.limit_denominator(3)
8.14 µs ± 172 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [15]: %timeit f2.limit_denominator(3)
1.84 µs ± 30.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


# int64 fractions show proportionally less speedup, but
# much larger time savings:
In [21]: f1 = F1(123_456_789_012_345, 234_567_890_123_456)  # reduced form
In [23]: f2 = F2(123_456_789_012_345, 234_567_890_123_456)

# large, but int32 denominator, 2.3x
In [35]: %timeit f1.limit_denominator(555_555_555)
12.1 µs ± 243 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [36]: %timeit f2.limit_denominator(555_555_555)
5.19 µs ± 94.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# int64 denominator, 1.9x
In [41]: %timeit f1.limit_denominator(5_555_555_555_555)
15.4 µs ± 283 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [42]: %timeit f2.limit_denominator(5_555_555_555_555)
7.92 µs ± 63.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


# BigInt with default limit_denominator  1.8x
In [47]: f1 = F1(284_467_440_737_095_516_100, 199_999_999_999_999_999_999)
In [49]: f2 = F2(284_467_440_737_095_516_100, 199_999_999_999_999_999_999)

In [50]: %timeit f1.limit_denominator()
15.2 µs ± 283 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
In [51]: %timeit f2.limit_denominator()
8.36 µs ± 176 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

```

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py:254]: We've lost some readability here: the original code clearly expressed the intent "if bound2 is closer to self than bound1 then return bound2, else return bound 1". With this code, I find it much harder to glean that intent.

A comment or two explaining what's going on might help.

I'd also suggest a bit of restructuring to keep things that are closely related together (and to keep consistency with the style of the rest of the function):
```suggestion
        bound1_n, bound1_d = p0 + k*p1, q0 + k*q1
        bound2_n, bound2_d = p1, q1
```
(And yes, there's some small overhead in the tuple packing and unpacking, but I think it's worth keeping for the readability.)

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: It's rare that I complain about names being too long, but I think these could be shortened. How about `diff1_n`, `diff1_d`, `diff2_n` and `diff2_d`? I'd also suggest dropping the redundant parentheses around the multiplications.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Suggest a more direct `if diff1_n * diff2_d >= diff2_n * diff1_d: ...`. (It also saves an operation.)

Comment (mscuthbert) [reply Lib/fractions.py]: Thanks @mdickinson  -- it's late here (Hawai'i) so will get this first thing in the morning.  Thanks for the close review.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: As you suggested in your comments, it's safe to use `_normalize=False` here and below: you can verify by induction that `p0 * q1 - p1 * q0` is `±1` at the beginning and the end of each iteration of the `while` loop (the sign alternates on each iteration), and it follows that `gcd(p0 + k*p1, q0 + k*q1) = 1` and that `gcd(p1, q1) = 1`.

**Review by mdickinson — CHANGES_REQUESTED** [review]
Thanks for the PR. The idea seems good, but we need to find a way to restore readability in the new code. I've left some in-line comments.

File: Lib/fractions.py
Comment (mdickinson) [reply Lib/fractions.py]: Yep, there's no hurry here.

**Comment by mdickinson** [comment]
> the general rule for all these methods is to trust that they are being called on a normalized Fraction only

Yes: there shouldn't be any way via the public API to end up with a non-normalised fraction, so methods can reasonably assume that they're operating on normalised fractions. If a `fractions` user is doing things like modifying the `_numerator` or `_denominator` fields of a fraction directly, then they do so at their own risk.

**Comment by mdickinson** [comment]
Just to be clear:

> We can eliminate the fourth gcd call also [...]

+1 We should definitely do this.

File: Lib/fractions.py
Comment (mscuthbert) [reply Lib/fractions.py]: That's what I thought too! but didn't want to suggest a change beyond the scope of the Issue/PR.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Does this change make a significant difference to performance? If not, I'd suggest reverting it, both for readability and for the purposes of keeping the PR focused. From looking at the code, a simple `Fraction(self)` shouldn't end up doing any gcd calculation here.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py:265]: I just realised that I'm not 100% convinced that the result here _is_ always normalised. There are two parts to being normalised: the numerator must be relatively prime to the denominator, and the denominator must be positive. I'd thought about the first part (which is fine), but not about the second.

Is it clear that `bound2_d` and `bound1_d` are positive at this point?

Comment (mdickinson) [reply Lib/fractions.py:265]: Answer: yes, it's true that `bound2_d` and `bound1_d` are positive at this point, though perhaps not 100% clear. In the Euclidean loop above, on the very _first_ iteration `a` can be negative, but on all subsequent iterations `a` is nonnegative. And on the first iteration `q1 = 0`, so that potentially negative `a` does no harm, and `q0` and `q1` are guaranteed nonnegative at all times. Then it's easy to check that they must in fact be positive at the point where the loop exits.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Let's drop this comment for this PR, and keep the PR focused on the `limit_denominator` changes. We could open a separate PR for this change.

File: Lib/fractions.py
Comment (mscuthbert) [reply Lib/fractions.py]: It actually saves 30% on this case based on tests above -- however, I hadn't realized at the time that the `return self` was executed in this branch and no normalization was done.  The speed increase is probably based on the `isinstance()` check, which just keeps getting faster and faster in subsequent versions, so it might eventually be the faster route.  I've removed the change.

If Fractions are eventually to be truly immutable and hashable, then `return self` is the correct response here, but that is probably backwards incompatible with real code where people change `_numerator` or `_denominator` after instantiation.

File: Lib/fractions.py
Comment (mscuthbert) [reply Lib/fractions.py:265]: That was my reading of the loop as well, but I didn't write the code (nor name the variables) so I didn't want to change this without another set of eyes; thanks!

**Comment by mscuthbert** [comment]
@mdickinson all comments addressed.  When merged I'll open a separate PR for the comment, since I do think it will be helpful for future maintainers to know what their obligations are or are not to non-normalized Fractions.

(NonNormalizedFraction would be a great superclass of Fraction at some point.  We used them in music a lot, where a 4/4 Time signature is not totally equivalent to 2/2 or 1/1 -- just as 2 and 2.0 aren't the same in many scientific domains.   I've had to make an entirely separate Rational subclass to work with them.  But another PR).

**Comment by mscuthbert** [comment]
I have made the requested changes; please review again

**Comment by rhettinger** [comment]
I concur with Mark that we value readability here more than a small speed boost.

This PR greatly obfuscates the code and provides only marginal benefit (this is the usual story when inlining code).

**Comment by mscuthbert** [comment]
okay -- a 3x speed boost for smallish numbers is literally why we can't use the base Fraction class in my project. I thought that I had been working to address @mdickinson's concerns, but I guess I had misread the tone of the conversation.

**Comment by rhettinger** [comment]
>  I guess I had misread the tone of the conversation.

It's possible that I'm the one who misread it.  If Mark wants to bring this back to life, that would be fine.  My view is that this isn't a method that needs optimizing and think the code should remain as is.  

ISTM that the entire point of the fractions module is to produce a class so that code won't have to handle numerators and denominators separately.  Presumably, any code that uses Fraction could be sped-up by inlining and not using the Fraction class, but that isn't what we've after.   We tend to save inlining for when we really need it.  IMO, the *limit_denominator()* method doesn't qualify.  It is almost never used in a performance critical loop, so it is not a worthy optimization target.

**Comment by mscuthbert** [comment]
> > I guess I had misread the tone of the conversation.
> 
> It's possible that I'm the one who misread it. If Mark wants to bring this back to life, that would be fine. My view is that this isn't a method that needs optimizing and think the code should remain as is.
> 
> ISTM that the entire point of the fractions module is to produce a class so that code won't have to handle numerators and denominators separately. Presumably, any code that uses Fraction could be sped-up by inlining and not using the Fraction class, but that isn't what we've after. We tend to save inlining for when we really need it. IMO, the _limit_denominator()_ method doesn't qualify. It is almost never used in a performance critical loop, so it is not a worthy optimization target.

@rhettinger -- what if the `def _sub(a, b)` call was amended to allow for a private `_normalize: bool` argument.  then it could be:

```
    def _sub(a, b, *, _normalize=True):
        """a - b"""
        na, da = a.numerator, a.denominator
        nb, db = b.numerator, b.denominator
        g = math.gcd(da, db) **if _normalize else 1**
        if g == 1:
            return Fraction(na * db - da * nb, da * db, _normalize=False)
        ....
```

which would allow for using nearly all the original code except that lines like `bound2 - self` would be `bound2._sub(self, _normalize=False)` and be very close to the original w/o inlining?  It would lose the constant-time speedups of this PR, but keep the O(log n) speedups that made the code much faster for most cases.

Libraries that mix float and Fraction (with denominators < a certain number) operations need to call limit_denominator often to know whether a float brought in from non-Python data such as 1/3 is intended to be interpreted as a fraction (my music21 library (600k yearly downloads) and at least two others I've found do this often), so for these cases speed of limit_denominator can be a limiting factor.  Implementing these code speedups alone moved the runtime for the music21 test suite from 8 minutes to 4 minutes.  I'd like give back to the Python community the result of a lot of trial and error for my lab at MIT so others don't need to find the same work arounds.  Thanks!

**Comment by mdickinson** [comment]
> I thought that I had been working to address @mdickinson's concerns

Indeed, that was my impression too.

**Comment by mdickinson** [comment]
I've broken out a minor non-invasive optimisation (simply avoiding unnecessarily normalisation in the upper and lower bounds) into #93730.

**Pull request closed without merging**
